### PR TITLE
feat: [CCM-8054]: Rule details page

### PR DESCRIPTION
**Pull request by rishik30**
#### Screenshots
Screenshot 2022-08-18 at 3 04 28 PM
Screenshot 2022-08-18 at 3 04 37 PM
Screenshot 2022-08-18 at 3 04 50 PM

Screenshot 2022-08-18 at 3 05 54 PM

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 201 Issues 
* Code issue 201 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/e864badd0b95ab81c060b15b85f15701/pull-request/10483)

**Comment by rishik30** [comment]
retrigger jest

**Comment by rishik30** [comment]
retrigger jest

File: src/modules/75-ce/common/CopyButton.tsx [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+  return (
+    <Button
+      minimal
+      icon="duplicate"
+      onClick={() => {
+        copy(textToCopy)
+        showCopySuccess()
+      }}
+      tooltip={'Copied'}
```
Comment (jyotiarora-harness) [inline src/modules/75-ce/common/CopyButton.tsx:37]: Use this text from string file

File: src/modules/75-ce/components/COGatewayList/COGatewayList.tsx [hunk 2]
```diff
@@ -1304,9 +1274,10 @@ const COGatewayList: React.FC = () => {
           rules={tableData}
           setRules={setTableData}
           loading={loading}
-          onRowClick={(e, index) => {
-            setSelectedService({ data: e, index })
-            setIsDrawerOpen(true)
+          onRowClick={e => {
+            // setSelectedService({ data: e, index })
+            // setIsDrawerOpen(true)
```
Comment (rajarshee-harness) [inline src/modules/75-ce/components/COGatewayList/COGatewayList.tsx:1279]: We should remove this if it's not gonna be used in future.

File: src/modules/75-ce/components/COGatewayList/Utils.tsx [hunk 3]
```diff
@@ -78,14 +73,8 @@ export function getInstancesLink(service: Service, resources?: AllResourcesOfAcc
...
-const gatewayStateMap: { [key: string]: JSX.Element } = {
-  down: <StoppedStatusIndicator />,
-  active: <RunningStatusIndicator />,
-  created: <CreatedStatusIndicator />
-}
-
-export function getStateTag(state: string): JSX.Element {
-  return gatewayStateMap[state]
+export function getStateTag(): JSX.Element {
+  return <></>
```
Comment (rajarshee-harness) [inline src/modules/75-ce/components/COGatewayList/Utils.tsx:77]: The function is just returning an empty fragment?

File: src/modules/75-ce/components/RuleDetails/RuleDetailsBody.tsx [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+  return (
+    <PageBody className={css.ruleDetailsBody}>
+      <Layout.Horizontal>
+        <RuleDataVisulisation service={service} />
+        <div className={css.colDivider} />
+        <Container className={css.col2}>
+          <Tabs id={'ruleDetailsTabs'}>
+            <Tab
+              id={'details'}
+              title="Details"
```
Comment (rajarshee-harness) [inline src/modules/75-ce/components/RuleDetails/RuleDetailsBody.tsx:128]: Can be moved to `strings.en.yaml`

File: src/modules/75-ce/components/RuleDetails/RuleDetailsBody.tsx [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+        (acc, curr) => acc + (curr.actual_savings as number),
+        0
+      ),
+      0
+    )
+  }, [savingsData?.response])
+
+  const totalActualCost = useMemo(() => {
+    return defaultTo(
+      defaultTo(get(savingsData, 'response') as ServiceSavings[], []).reduce(
```
Comment (rajarshee-harness) [inline src/modules/75-ce/components/RuleDetails/RuleDetailsBody.tsx:197]: I think `get` already takes a third parameter for default value.

File: src/modules/75-ce/components/RuleDetails/RuleDetailsBody.tsx [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+                  accessPointData={get(accessPointData, 'response')}
+                  resources={get(resources, 'response')}
+                  dependencies={dependencies}
+                  setService={setService}
+                />
+              }
+            />
+            <Tab
+              id={'ssh'}
+              title="SSH/RDP via Harness CLI "
```
Comment (jyotiarora-harness) [inline src/modules/75-ce/components/RuleDetails/RuleDetailsBody.tsx:144]: We can use this from strings

File: src/modules/75-ce/components/RuleDetails/RuleStatusToggleSwitch.tsx [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  const handleToggleClick = () => {
+    setLoading(true)
+    triggerToggle()
+  }
+
+  return (
+    <Layout.Horizontal>
+      <Popover
+        position={Position.TOP}
+        content={<Text padding={'small'}>{'Enable/Disable text here'}</Text>}
```
Comment (rajarshee-harness) [inline src/modules/75-ce/components/RuleDetails/RuleStatusToggleSwitch.tsx:63]: Can be moved to `strings.en.yaml`

File: src/modules/75-ce/components/RuleDetails/RuleDetailsTabContainer.tsx [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+                Cell: ({ row }) => (
+                  <Layout.Vertical>
+                    {getTableText(row.original.name, { font: { variation: FontVariation.SMALL_SEMI } })}
+                    {getTableText(row.original.id, { color: Color.GREY_400 })}
+                  </Layout.Vertical>
+                )
+              },
+              {
+                accessor: 'ipv4',
+                Header: getTableText('Public/Private IP', { font: { variation: FontVariation.SMALL_BOLD } }),
```
Comment (jyotiarora-harness) [inline src/modules/75-ce/components/RuleDetails/RuleDetailsTabContainer.tsx:145]: Should be driven from strings

File: src/modules/75-ce/components/RuleDetails/RuleDetailsTabContainer.tsx [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+                Cell: ({ row }) => (
+                  <Layout.Vertical>
+                    {getTableText(get(row.original, 'ipv4.0'), { font: { variation: FontVariation.SMALL_SEMI } })}
+                    {getTableText(get(row.original, 'private_ipv4.0'), { color: Color.GREY_400 })}
+                  </Layout.Vertical>
+                )
+              },
+              {
+                accessor: 'type',
+                Header: getTableText('Instance family', { font: { variation: FontVariation.SMALL_BOLD } }),
```
Comment (jyotiarora-harness) [inline src/modules/75-ce/components/RuleDetails/RuleDetailsTabContainer.tsx:156]: Should be driven from strings

File: src/modules/75-ce/components/RuleDetails/RuleDetailsTabContainer.tsx [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+                width: '25%',
+                Cell: ({ row }) => (
+                  <Layout.Vertical>
+                    {getTableText(row.original.type, { font: { variation: FontVariation.SMALL_SEMI } })}
+                  </Layout.Vertical>
+                )
+              },
+              {
+                accessor: 'tags',
+                Header: getTableText('Tags', { font: { variation: FontVariation.SMALL_BOLD } }),
```
Comment (jyotiarora-harness) [inline src/modules/75-ce/components/RuleDetails/RuleDetailsTabContainer.tsx:166]: Same as above

File: src/modules/75-ce/components/RuleDetails/RuleDetailsBody.tsx (hunk 5 above)
Comment (rishik30) [reply src/modules/75-ce/components/RuleDetails/RuleDetailsBody.tsx:197]: yes it does but if the key has null value, then it doesn't fallback to default value. It takes default only when the key is not present.

File: src/modules/75-ce/components/COGatewayList/Utils.tsx (hunk 3 above)
Comment (rishik30) [reply src/modules/75-ce/components/COGatewayList/Utils.tsx:77]: ya there is a clean up required which I will take up in a follow-up PR

File: src/modules/75-ce/pages/co-rule-details/CORuleDetailsPage.tsx [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+                  <Text font={{ variation: FontVariation.H4 }}>{defaultTo(service?.name, '')}</Text>
+                  {service && (
+                    <Container>
+                      <RuleStatusToggleSwitch serviceData={service} onSuccess={setService} />
+                    </Container>
+                  )}
+                </Layout.Horizontal>
+                <Layout.Horizontal spacing={'small'} margin={{ top: 'small' }}>
+                  <Text color={Color.GREY_500} font={{ variation: FontVariation.BODY, size: 'small' }}>
+                    {'AWS EC2'}
```
Comment (jyotiarora-harness) [inline src/modules/75-ce/pages/co-rule-details/CORuleDetailsPage.tsx:153]: Please use this from string file

File: src/modules/75-ce/common/CopyButton.tsx (hunk 1 above)
Comment (rishik30) [reply src/modules/75-ce/common/CopyButton.tsx:37]: done

File: src/modules/75-ce/components/RuleDetails/RuleDetailsBody.tsx (hunk 6 above)
Comment (rishik30) [reply src/modules/75-ce/components/RuleDetails/RuleDetailsBody.tsx:144]: done

File: src/modules/75-ce/components/COGatewayList/COGatewayList.tsx (hunk 2 above)
Comment (rishik30) [reply src/modules/75-ce/components/COGatewayList/COGatewayList.tsx:1279]: done

File: src/modules/75-ce/components/RuleDetails/RuleDetailsBody.tsx (hunk 4 above)
Comment (rishik30) [reply src/modules/75-ce/components/RuleDetails/RuleDetailsBody.tsx:128]: done

File: src/modules/75-ce/components/RuleDetails/RuleStatusToggleSwitch.tsx (hunk 7 above)
Comment (rishik30) [reply src/modules/75-ce/components/RuleDetails/RuleStatusToggleSwitch.tsx:63]: done

File: src/modules/75-ce/components/RuleDetails/RuleDetailsTabContainer.tsx (hunk 9 above)
Comment (rishik30) [reply src/modules/75-ce/components/RuleDetails/RuleDetailsTabContainer.tsx:156]: done

File: src/modules/75-ce/components/RuleDetails/RuleDetailsTabContainer.tsx (hunk 10 above)
Comment (rishik30) [reply src/modules/75-ce/components/RuleDetails/RuleDetailsTabContainer.tsx:166]: done

File: src/modules/75-ce/pages/co-rule-details/CORuleDetailsPage.tsx (hunk 11 above)
Comment (rishik30) [reply src/modules/75-ce/pages/co-rule-details/CORuleDetailsPage.tsx:153]: done

**Comment by rishik30** [comment]
trigger cypress

**Comment by rishik30** [comment]
trigger cypress

File: src/modules/75-ce/components/COGatewayList/COGatewayList.tsx [hunk 12]
```diff
@@ -1304,9 +1274,8 @@ const COGatewayList: React.FC = () => {
           rules={tableData}
           setRules={setTableData}
           loading={loading}
-          onRowClick={(e, index) => {
-            setSelectedService({ data: e, index })
-            setIsDrawerOpen(true)
+          onRowClick={e => {
+            history.push(routes.toCECORuleDetails({ accountId, id: e.id?.toString() as string }))
```
Comment (jyotiarora-harness) [inline src/modules/75-ce/components/COGatewayList/COGatewayList.tsx:1278]: toString will typecast it to string, then do we need to define the type as string?

Comment (rishik30) [reply src/modules/75-ce/components/COGatewayList/COGatewayList.tsx:1278]: we need to as the id is an optional param

**Review by rajarshee-harness — APPROVED** [review]

File: src/modules/75-ce/components/COGatewayList/COGatewayList.tsx (hunk 12 above)
Comment (rajarshee-harness) [reply src/modules/75-ce/components/COGatewayList/COGatewayList.tsx:1278]: Maybe use `String()`

**Comment by rishik30** [comment]
retrigger jest

**Review by rajarshee-harness — APPROVED** [review]

**Review by jyotiarora-harness — APPROVED** [review]

**Comment by rishik30** [comment]
retrigger jest

**Review by kapilkumawat86 — APPROVED** [review]

**Pull request merged**